File: SDThemer.swift
import SwiftUI

struct SDThemer {
    static var theme: SDTheme = SDThemes.default.sdTheme
}

protocol SDTheme {
    var strings: any SDStrings { get }
    var dimensions: any SDDimensions { get }
    var colors: any SDColors { get }
    var fonts: any SDFonts { get }
}

protocol SDStrings {
    var label: String { get }
}

protocol SDDimensions {
    var width: CGFloat { get }
    var height: CGFloat { get }
}

protocol SDColors {
    var primary: Color { get }
    var secondary: Color { get }
}

protocol SDFonts {
    var body: Font { get }
}

enum SDThemes: CaseIterable {
    case custom
    case `default`

    var sdTheme: SDTheme {
        switch self {
        case .custom: return CustomSDTheme()
        case .`default`: return DefaultSDTheme()
        }
    }
}

private extension UIColor {
    convenience init(hex: String) {
        var hexString = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if hexString.hasPrefix("#") { hexString.removeFirst() }
        var rgbValue: UInt64 = 0
        Scanner(string: hexString).scanHexInt64(&rgbValue)
        self.init(
            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
            alpha: 1.0
        )
    }
}

struct CustomSDTheme: SDTheme {

    var strings: any SDStrings = Strings()
    var dimensions: any SDDimensions = Dimensions()
    var colors: any SDColors = Colors()
    var fonts: any SDFonts = Fonts()

    struct Strings: SDStrings {
        var label: String { return "" }
    }

    struct Dimensions: SDDimensions {
        var width: CGFloat { 0 }
        var height: CGFloat { 0 }
    }

    struct Colors: SDColors {
        var primary: Color {
            return Color(UIColor { traitCollection in
                switch traitCollection.userInterfaceStyle {
                case .dark:
                    return UIColor(hex: "#FFFFFF")
                default:
                    return UIColor(hex: "#000000")
                }
            })
        }
        var secondary: Color { return Color(UIColor(hex: "#808080")) }
    }

    struct Fonts: SDFonts {
        var body: Font { .custom("", size: 0) }
    }
}

struct DefaultSDTheme: SDTheme {

    var strings: any SDStrings = Strings()
    var dimensions: any SDDimensions = Dimensions()
    var colors: any SDColors = Colors()
    var fonts: any SDFonts = Fonts()

    struct Strings: SDStrings {
        var label: String { return "" }
    }

    struct Dimensions: SDDimensions {
        var width: CGFloat { 0 }
        var height: CGFloat { 0 }
    }

    struct Colors: SDColors {
        var primary: Color {
            return Color(UIColor { traitCollection in
                switch traitCollection.userInterfaceStyle {
                case .dark:
                    return UIColor(hex: "#FFFFFF")
                default:
                    return UIColor(hex: "#000000")
                }
            })
        }
        var secondary: Color { return Color(UIColor(hex: "#808080")) }
    }

    struct Fonts: SDFonts {
        var body: Font { .custom("", size: 0) }
    }
}

File: SDThemer.kt
package com.example.theme
import com.example.theme.R

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import android.graphics.Color as AndroidColor

object SDThemer {
    var theme: SDTheme = SDThemes.DEFAULT.sdTheme()
}

interface SDTheme {
    val strings: SDStrings
    val dimensions: SDDimensions
    val bools: SDBools
    val colors: SDColors
    val fonts: SDFonts
}
  
interface SDStrings {
    val label: String
}
  
interface SDDimensions {
    val width: Float
    val height: Float
}
  
interface SDBools {
    val show: Boolean
}
  
interface SDColors {
    @Composable
    fun primary(): Color
    val secondary: Color
}
  
interface SDFonts {
      val body: TextStyle
}

enum class SDThemes {
    CUSTOM,
    DEFAULT;

    fun sdTheme(): SDTheme {
        return when (this) {
            CUSTOM -> CustomSDTheme
            DEFAULT -> DefaultSDTheme
        }
    }
}

object CustomSDTheme : SDTheme {

    override val strings: SDStrings = Strings
    override val dimensions: SDDimensions = Dimensions
    override val bools: SDBools = Bools
    override val colors: SDColors = Colors
    override val fonts: SDFonts = Fonts

    object Strings : SDStrings {
        override val label: String get() = ""
    }

    object Dimensions : SDDimensions {
        override val width: Float get() = 0f
        override val height: Float get() = 0f
    }

    object Bools : SDBools {
        override val show: Boolean get() = false
    }

    object Colors : SDColors {
        @Composable
        override fun primary(): Color {
            return if (isSystemInDarkTheme()) {
                Color(AndroidColor.parseColor("#FFFFFF"))
            } else {
                Color(AndroidColor.parseColor("#000000"))
            }
        }
        override val secondary: Color = Color(AndroidColor.parseColor("#808080"))
    }

    object Fonts : SDFonts {
        override val body: TextStyle get() = TextStyle(
            fontFamily = FontFamily(Font(R.font.Arial)),
            fontSize = 0.sp
        )
    }
}


object DefaultSDTheme : SDTheme {

    override val strings: SDStrings = Strings
    override val dimensions: SDDimensions = Dimensions
    override val bools: SDBools = Bools
    override val colors: SDColors = Colors
    override val fonts: SDFonts = Fonts

    object Strings : SDStrings {
        override val label: String get() = ""
    }

    object Dimensions : SDDimensions {
        override val width: Float get() = 0f
        override val height: Float get() = 0f
    }

    object Bools : SDBools {
        override val show: Boolean get() = false
    }

    object Colors : SDColors {
        @Composable
        override fun primary(): Color {
            return if (isSystemInDarkTheme()) {
                Color(AndroidColor.parseColor("#FFFFFF"))
            } else {
                Color(AndroidColor.parseColor("#000000"))
            }
        }
        override val secondary: Color = Color(AndroidColor.parseColor("#808080"))
    }

    object Fonts : SDFonts {
        override val body: TextStyle get() = TextStyle(
            fontFamily = FontFamily(Font(R.font.Arial)),
            fontSize = 0.sp
        )
    }
}
